File: react-kanban/src/types/modals/board.ts
import { useState } from "react";
import type { Board } from "../serialize";
import type { BoardValidationError, UserBoard } from "../validation";

export type OpenBoardModal = {
  board: Board | null;
};

export function openBoardModal(): OpenBoardModal {
  return { board: null };
}

export function openBoardModalWithBoard(board: Board): OpenBoardModal {
  return { board };
}

export function useBoardForm(
  board: Board | null,
  modalData: OpenBoardModal | null,
) {
  const [formModalData, setFormModalData] = useState(modalData);
  const [title, setTitle] = useState(board?.title ?? "");
  const [titleError, setTitleError] = useState<string>();

  if (modalData !== formModalData) {
    setFormModalData(modalData);
    setTitle(board?.title ?? "");
    setTitleError(undefined);
  }

  const userBoard = (): UserBoard => ({ title });
  const clearError = () => setTitleError(undefined);
  const setValidationError = (error: BoardValidationError) => {
    setTitleError(
      error === "EmptyTitle"
        ? "The title must not be empty"
        : "Please enter a title with at most 200 characters.",
    );
  };

  return {
    title,
    setTitle,
    titleError,
    userBoard,
    clearError,
    setValidationError,
  };
}

export type BoardForm = ReturnType<typeof useBoardForm>;

File: leptos-kanban/src/types/modals/board.rs
use leptos::prelude::*;

use crate::types::{
    state::BoardState,
    validation::board::{BoardValidationError, UserBoard},
};

#[derive(Clone)]
pub struct OpenBoardModal {
    pub board: Option<BoardState>,
}

impl OpenBoardModal {
    pub fn new() -> Self {
        Self { board: None }
    }

    pub fn new_with_board(board: BoardState) -> Self {
        Self { board: Some(board) }
    }
}

#[derive(Clone, Copy)]
pub struct BoardFormState {
    pub title: RwSignal<String>,
    pub title_error: RwSignal<Option<String>>,
}

impl BoardFormState {
    pub fn new(board: Option<BoardState>) -> Self {
        let form = Self {
            title: RwSignal::new(String::new()),
            title_error: RwSignal::new(None),
        };
        form.set_board(board);
        form
    }

    pub fn set_board(self, board: Option<BoardState>) {
        self.title.set(
            board
                .map(|board| board.title.get_untracked())
                .unwrap_or_default(),
        );
        self.clear_error();
    }

    pub fn user_board(self) -> UserBoard {
        UserBoard {
            title: self.title.get_untracked(),
        }
    }

    pub fn clear_error(self) {
        self.title_error.set(None);
    }

    pub fn set_validation_error(self, error: BoardValidationError) {
        let message = match error {
            BoardValidationError::EmptyTitle => "The title must not be empty",
            BoardValidationError::TitleTooLong => {
                "Please enter a title with at most 200 characters."
            }
        };

        self.title_error.set(Some(String::from(message)));
    }
}
